src/hostapi/pulseaudio/pa_hostapi_pulseaudio_cb.c

```diff
@@ -0,0 +1,483 @@
+    pa_operation *l_ptrOperation = NULL;
+
+    pa_threaded_mainloop_wait(l_ptrPulseaudioHostApi->mainloop);
+
+    if(stream->outStream != NULL)
+    {
+        /* First we stop (CORK) stream and make sure
+         * that we don't write anymore
+         */
+        l_ptrOperation = pa_stream_cork(stream->outStream, 1, PulseaudioStreamSuccessCb, NULL);
+
+        while (pa_operation_get_state(l_ptrOperation) == PA_OPERATION_RUNNING)
+        {
+            pa_threaded_mainloop_wait(l_ptrPulseaudioHostApi->mainloop);
+        }
+
+        pa_operation_unref(l_ptrOperation);
+
+
+        /* Then we cancel all writing (if there is any)
+         *  and wait for disconnetion (TERMINATION) which
+         * can take a while for ethernet connections
+         */
+        pa_stream_cancel_write(stream->outStream);
+        pa_stream_disconnect(stream->outStream);
+
+
+        while(pa_stream_get_state(stream->outStream) != PA_STREAM_TERMINATED)
+        {
+            pa_threaded_mainloop_wait(l_ptrPulseaudioHostApi->mainloop);
+        }
+
+        pa_stream_unref(stream->outStream);
+        stream->outStream = NULL;
+
+        PaUtil_FreeMemory(stream->outBuffer);
+        stream->outBuffer = NULL;
+    }
+
+    if(stream->inStream != NULL)
+    {
+        /* First we stop (CORK) stream and make sure
+         * that we don't read anymore
+         */
+        l_ptrOperation = pa_stream_cork(stream->inStream, 1, PulseaudioStreamSuccessCb, NULL);
+
+        while (pa_operation_get_state(l_ptrOperation) == PA_OPERATION_RUNNING)
+        {
+            pa_threaded_mainloop_wait(l_ptrPulseaudioHostApi->mainloop);
+        }
+
+        pa_operation_unref(l_ptrOperation);
+
+        /* Then we disconnect stream and wait for
+         * Termination
+         */
+        pa_stream_disconnect(stream->inStream);
+
+        while(pa_stream_get_state(stream->inStream) != PA_STREAM_TERMINATED)
+        {
+            pa_threaded_mainloop_wait(l_ptrPulseaudioHostApi->mainloop);
+        }
+
+        pa_stream_unref(stream->inStream);
+        stream->inStream = NULL;
+
+        PaUtil_FreeMemory(stream->inBuffer);
+        stream->inBuffer = NULL;
+    }
+
+    PaUtil_TerminateBufferProcessor(&stream->bufferProcessor);
+    PaUtil_TerminateStreamRepresentation(&stream->streamRepresentation);
+    PaUtil_FreeMemory(stream);
+
+    stream->isStopped = 1;
+    stream->isActive = 0;
+
+    return result;
+}
+
+
+PaError PulseaudioStartStreamCb(PaStream *s)
+{
+    PaError result = paNoError;
+    PaPulseaudioStream *stream = (PaPulseaudioStream*)s;
+    int streamStarted = 0;  /* So we can know whether we need to take the stream down */
+    PaPulseaudioHostApiRepresentation *l_ptrPulseaudioHostApi = stream->hostapi;
+    const char *l_strName = NULL;
+
+    pa_threaded_mainloop_lock(l_ptrPulseaudioHostApi->mainloop);
+
+    /* Ready the processor */
+    PaUtil_ResetBufferProcessor(&stream->bufferProcessor);
+
+    stream->latency = 20000;
```

Comment:
I don't really remember. I think it felt nice in the days I implemented this :rabbit: